ta/pkcs11/src/attributes.c

```diff
@@ -0,0 +1,299 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2017-2020, Linaro Limited
+ */
+
+#include <assert.h>
+#include <compiler.h>
+#include <pkcs11_ta.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+#include <string_ext.h>
+#include <tee_internal_api.h>
+#include <tee_internal_api_extensions.h>
+#include <trace.h>
+#include <util.h>
+
+#include "attributes.h"
+#include "pkcs11_helpers.h"
+#include "serializer.h"
+
+enum pkcs11_rc init_attributes_head(struct obj_attrs **head)
+{
+	*head = TEE_Malloc(sizeof(**head), TEE_MALLOC_FILL_ZERO);
+	if (!*head)
+		return PKCS11_CKR_DEVICE_MEMORY;
+
+	return PKCS11_CKR_OK;
+}
+
+enum pkcs11_rc add_attribute(struct obj_attrs **head, uint32_t attribute,
+			     void *data, size_t size)
+{
+	size_t buf_len = sizeof(struct obj_attrs) + (*head)->attrs_size;
+	char **bstart = (void *)head;
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+	uint32_t data32 = 0;
+
+	data32 = attribute;
+	rc = serialize(bstart, &buf_len, &data32, sizeof(uint32_t));
+	if (rc)
+		return rc;
+
+	data32 = size;
+	rc = serialize(bstart, &buf_len, &data32, sizeof(uint32_t));
+	if (rc)
+		return rc;
```

Comment:
That's not needed and will not change anything either. `head` is a pointer to a pointer `*head` is always holding the same value pointer value as `*bstart` since they are aliases of each other.

Reply:
yes, you're right. sorry for the noise.